### PR TITLE
Do not fire multiple events

### LINKED ISSUE #6032: Setting derived properties in evented models causes duplicate events
## 🐛 Bug

Changing the value of a settable derived property in an `EventedModel` causes the corresponding event to be executed twice. For example, `Dims.current_step` is derived from `Dims.point`, but can also be set. Changing its value causes `Dims.events.current_step` to be executed twice.

## To Reproduce

Here is a general example with a custom evented model.

```python
from napari.utils.events import EventedModel

class Model(EventedModel):
    a: int

    @property
    def b(self) -> int:
        return self.a + 1

    @b.setter
    def b(self, b) -> None:
        self.a = b - 1

m = Model(a=0)
m.events.a.connect(lambda _: print('a'))
m.events.b.connect(lambda _: print('b'))
m.a = 1 # prints "a" and "b" once each, as expected
m.b = 3 # prints "a" once, but "b" twice
```

Here's a specific example with `Dims.current_step`.

```python
from napari.components import Dims
d = Dims()
d.events.current_step.connect(lambda _: print('current_step'))
d.current_step = (1, 0)
# prints "current_step" twice
```

## Expected behavior

The event should be executed once.

## Additional context

While this bug exists, `ViewerModel` connects `current_step` to `_update_layers`, which may cause slicing to occur twice in some cases (though early returns may save us). We might want to use the `point` event instead, which is only executed once when `point` or `current_step` is set.

### PR DESCRIPTION
**Pull request by brisvag**
# Fixes/Closes

Closes #6032

# Description

As suggested by @andy-sweet, block firing of events by caused by changes in dependent properties to avoid duplicate events.

Tests were failing due to mocks not having context manager behaviour, but this seems be solved by a simple subclass.

### CODE REVIEW

File: napari/utils/events/_tests/test_evented_model.py
Comment (andy-sweet) [inline napari/utils/events/_tests/test_evented_model.py]: Optional: I'm not too familiar with Python's mocks, but maybe we could just use [`MagicMock`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.MagicMock)? Might be a little slower, but I suspect it's negligible.

File: napari/utils/events/_tests/test_evented_model.py
Comment (andy-sweet) [inline napari/utils/events/_tests/test_evented_model.py]: If we keep this, then maybe call it `EventEmitterMock` or `EmitterMock` as this mocks `EventEmitter`, right?

File: napari/utils/events/_tests/test_evented_model.py
Comment (andy-sweet) [inline napari/utils/events/_tests/test_evented_model.py]: I might be still waking up, but how does mock `EventEmitter.blocker` as desired?

File: napari/utils/events/_tests/test_evented_model.py
Comment (andy-sweet) [inline napari/utils/events/_tests/test_evented_model.py]: Optional: I think using a plain `Mock` with `assert_called_once` as done elsewhere in these tests is fine and doesn't require defining this extra code. But I don't feel strongly here.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Pretty much looks good to me. Only requested changes before approval are answers to questions and clarifying comments. Also left some other suggestions/ideas, but those are optional.

File: napari/utils/events/_tests/test_evented_model.py
Comment (brisvag) [reply napari/utils/events/_tests/test_evented_model.py]: I swear I tried and it failed... now it seems to work :facepalm: switching!

File: napari/utils/events/_tests/test_evented_model.py
Comment (brisvag) [reply napari/utils/events/_tests/test_evented_model.py]: It does not; this is only needed to avoud *errors* in the previous tests, but it does not ensure correct behaviour for this specific thing. This is why the new test actually uses an ad-hoc system. Ultimately I think we should either replace everything with a proper system, or subclass `Mock` properly somehow.

File: napari/utils/events/_tests/test_evented_model.py
Comment (brisvag) [reply napari/utils/events/_tests/test_evented_model.py]: Mentioned above, but I'll expand here: the behaviour relies on the `blocker` context manager, which break down with `Mock` objects. I could not find a way around it by just subclassing or using the magicmock. If you do, please let me know! I find this implementation actually a bit clearer and less verbose than all the mock stuff, so I don't mind this personally.

Comment (andy-sweet) [reply napari/utils/events/_tests/test_evented_model.py]: I meant using the `Mock` as the callback, as in `test_events_are_fired_only_if_necessary`

```python
    t = Tt()
    a_callback = Mock()
    b_callback = Mock()
    t.events.a.connect(a_callback)
    t.events.b.connect(b_callback)
    t.a = 2
    a_callback.assert_called_once()
    b_callback.assert_called_once()
```

Still don't feel strongly here, but I'd prefer to have one consistent way to doing this within at least this test module.

I wondered if we could modify `test_evented_model_with_property_setter_events` instead of adding a new test, but I haven't grokked that properly yet.

I think it's also OK to keep this, so approving and I'll let you decide.

File: napari/utils/events/_tests/test_evented_model.py
Comment (andy-sweet) [inline napari/utils/events/_tests/test_evented_model.py]: I assume this meant `Mock` as a verb and this just got updated by a search/replace?

```suggestion
    # Mock events
```

I don't think I can stomach `MagicMock` as a verb :p

Alternatively, consider deleting the comment because it adds very little here - I already clearly see that we're mocking events (or rather the `values` emitter) from the code itself.

**Review by andy-sweet — APPROVED** [review]
Looks good! I'm going to file a similar issue on psygnal if it's also present there. Also makes me think about whether we should consider depending on that instead, but that's a much bigger discussion.

**Review by Czaki — COMMENTED** [review]
If I correctly understand if we have class:

```python
class Tt(EventedModel):
    a: int = 1

    @property
    def b(self) -> float:
        return self.a * 2

    @b.setter
    def b(self, v):
        self.a = v // 2

    @property
    def c(self) -> float:
        return self.a * 3

    @c.setter
    def c(self, v):
        self.a = v // 3
```

Then calling 

```
obj.b = 2
```

will trigger events for `a` and `b` but not for `c`?

Maybe we should omit emitting events for setting a property as the base argument setting will fire all required events?

File: napari/utils/events/_tests/test_evented_model.py
Comment (brisvag) [reply napari/utils/events/_tests/test_evented_model.py]: lol, yup, you got me :P

**Comment by brisvag** [comment]
I'm not followin @Czaki... if you do `obj.b = 2` there, no events are fired cause nothing changes (`obj.a` is already `2 // 1`).

If you do `obj.b = 5` for example, everything works as intended (all 3 events are fired, and only once).

File: napari/utils/events/_tests/test_evented_model.py
Comment (brisvag) [reply napari/utils/events/_tests/test_evented_model.py]: Ok, switched to a mocked callback, much cleaner indeed :+1:

**Comment by Czaki** [comment]
I need to read code for creating deps dict again, but could you add such indirect deps to the test?

**Comment by andy-sweet** [comment]
> If I correctly understand if we have class:
> 
> ```python
> class Tt(EventedModel):
>     a: int = 1
> 
>     @property
>     def b(self) -> float:
>         return self.a * 2
> 
>     @b.setter
>     def b(self, v):
>         self.a = v // 2
> 
>     @property
>     def c(self) -> float:
>         return self.a * 3
> 
>     @c.setter
>     def c(self, v):
>         self.a = v // 3
> ```
> 
> Then calling
> 
> ```
> obj.b = 2
> ```
> 
> will trigger events for `a` and `b` but not for `c`?

@Czaki : good thinking about transitive dependencies. Extending the test to account seems like a good idea.

From some quick testing, before this PR (i.e. on `main`), I get one event for `a` and `c`, but two for `b`. After this PR I get one event for each, which seems like the desired outcome and is at least more desirable than `main`.

> Maybe we should omit emitting events for setting a property as the base argument setting will fire all required events?

Not sure what you mean here. In the example above, are you suggesting that only `a` should be an emitted event?

**Comment by Czaki** [comment]
I finally understand which event is blocked. So it looks like problem is partially solved and it is still possible to have multiple events. 

```python
class Tt(EventedModel):
    a: int = 1
    d: int = 0

    @property
    def c(self) -> float:
        return self.a + self.b

    @c.setter
    def c(self, v):
        self.a = v // 2
        self.b = v - self.a

    @property
    def d(self) -> float:
        return self.a + 2 * self.b

    @d.setter
    def d(self, v):
        self.a = v // 3
        self.b = (v - self.a) //2
```

and 

```python
obj.c = 7
```
Then I think that `d` will be emitted twice.

The main disadvantage of this solution for me is that cmp operator on `b` (from test) will be called twice. 

WHy would we like to perform a comparison when setting property? Why can events not be triggered by setting object attributes?

**Comment by andy-sweet** [comment]
> ```python
> ```python
> class Tt(EventedModel):
>     a: int = 1
>     d: int = 0
> 
>     @property
>     def c(self) -> float:
>         return self.a + self.b
> 
>     @c.setter
>     def c(self, v):
>         self.a = v // 2
>         self.b = v - self.a
> 
>     @property
>     def d(self) -> float:
>         return self.a + 2 * self.b
> 
>     @d.setter
>     def d(self, v):
>         self.a = v // 3
>         self.b = (v - self.a) //2
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> and
> ```python
> obj.c = 7
> ```
> 
> Then I think that d will be emitted twice.
> The main disadvantage of this solution for me is that cmp operator on b (from test) will be called twice.
> ```

Good point. And agreed the `d` event is emitted 2 times on this branch and all the other events are emitted 2 times. On `main`, `d` is also emitted 2 times, but `c` is emitted 3 times! So this PR represents a substantial improvement over `main`, even though there are still complex cases it can't quite handle.

Definitely open to more complete solutions.

> WHy would we like to perform a comparison when setting property? Why can events not be triggered by setting object attributes?

I can't answer this too well, but I think some related motivation is captured in https://github.com/napari/napari/pull/3711, https://github.com/napari/napari/pull/3586, and on [a zulip discussion](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/pydantic.20model.20with.20settable.20property).

From usage of the API, attributes and settable properties are indistinguishable. So my guess is that derived properties have events so that users don't have to understand the implementation details to understand what to connect to. I don't feel strongly about this (e.g. in the case where this caused an issue, I think I can just connect to the `point` event instead of `current_step`), but removing it would be a change in behavior.

**Comment by Czaki** [comment]
I have created #6051 with an alternative implementation. Maybe it could be better or some fusion. The #6051 address also problem of multiple emit of different properties.

**Comment by brisvag** [comment]
Closing in favour of #6051.

**Pull request closed without merging**
